protocols/autonatv2/src/request_response.rs

```diff
@@ -0,0 +1,336 @@
+// change to quick-protobuf-codec
+
+use std::io::ErrorKind;
+use std::{borrow::Cow, io};
+
+use asynchronous_codec::{Framed, FramedRead, FramedWrite};
+
+use futures::{AsyncRead, AsyncWrite, AsyncWriteExt, SinkExt, StreamExt};
+use libp2p_core::Multiaddr;
+
+use quick_protobuf_codec::Codec;
+use rand::Rng;
+
+use crate::{generated::structs as proto, Nonce};
+
+const REQUEST_MAX_SIZE: usize = 4104;
+pub(super) const DATA_LEN_LOWER_BOUND: usize = 30_000u32 as usize;
+pub(super) const DATA_LEN_UPPER_BOUND: usize = 100_000u32 as usize;
+pub(super) const DATA_FIELD_LEN_UPPER_BOUND: usize = 4096;
+
+macro_rules! new_io_invalid_data_err {
+    ($msg:expr) => {
+        io::Error::new(io::ErrorKind::InvalidData, $msg)
+    };
+}
+
+macro_rules! check_existence {
+    ($field:ident) => {
+        $field.ok_or_else(|| new_io_invalid_data_err!(concat!(stringify!($field), " is missing")))
+    };
+}
+
+pub(crate) struct Coder<I> {
+    inner: Framed<I, Codec<proto::Message>>,
+}
+
+impl<I> Coder<I>
+where
+    I: AsyncWrite + AsyncRead + Unpin,
+{
+    pub(crate) fn new(io: I) -> Self {
+        Self {
+            inner: Framed::new(io, Codec::new(REQUEST_MAX_SIZE)),
+        }
+    }
+    pub(crate) async fn close(self) -> io::Result<()> {
+        let mut stream = self.inner.into_inner();
+        stream.close().await?;
+        Ok(())
+    }
+}
+
+impl<I> Coder<I>
+where
+    I: AsyncRead + Unpin,
+{
+    async fn next_msg(&mut self) -> io::Result<proto::Message> {
+        self.inner
+            .next()
+            .await
+            .ok_or(io::Error::new(ErrorKind::Other, "no request to read"))?
+            .map_err(|e| io::Error::new(ErrorKind::InvalidData, e))
+    }
+    pub(crate) async fn next_request(&mut self) -> io::Result<Request> {
+        Request::from_proto(self.next_msg().await?)
+    }
+    pub(crate) async fn next_response(&mut self) -> io::Result<Response> {
+        Response::from_proto(self.next_msg().await?)
+    }
```

Comment:
You might like this idea:

- Implement `TryFrom` and `Into` to go between `Request` / `Response` and `proto::Message`
- Add the following implementations to `Coder`:

```rust
impl Coder<I> {
	pub(crate) async fn next<M>(&mut self) -> io::Result<M> where proto::Message: TryInto<M> {
		let msg = self.next_msg().await?.try_into()?;

		Ok(msg)
	}

	pub(crate) async fn send<M>(&mut self, msg: M) -> io::Result<()> where M: Into<proto::Message> {
		self.inner.send(msg.into()).await?;

		Ok(())
	}
}
```

Reply:
That's very elegant indeed